### DISCUSSION TITLE
useBaseQuery onSubscribe is slow when mounting lots of useQuery instances

### DISCUSSION THREAD

**Discussion by dwilt** [TanStack/query#3756] [Q&A] [answered]
First off: we have a pretty intense React application. We use `useMemo`, `useCallback` and `React.memo` pretty intensely to reduce re-renders. And these have had _huge_ wins from a performance perspective. We recently just migrated from `useSWR` to RQ because it handles infinite scrolling better. However, one thing I'm seeing that is a bit concerning is the overall performance of the application seems significantly worse. We're using [`notifyOnChangeProps: 'tracked'`](https://tkdodo.eu/blog/react-query-render-optimizations) to reduce re-renders which has been important to keep parity with how `useSWR` tracks which props are used. I'm not sure if we're missing some sort of config option but given that we have so many components on one page which utilize some query hook, whenever we change pages, there seem to be _hundreds_ of RQ hooks which mount. And using the performance profiler in Chrome, everything is pointing to `useBaseQuery`'s `onSubscribe` call. 

Take a look below at me navigating between pages in our application. You can see a bit of lag in the transition.
![2022-06-27 20 56 27](https://user-images.githubusercontent.com/1065539/176064102-98f2f244-d213-4944-ab57-d265134dfc7b.gif)

Again, it's worth noting that there are _hundreds_ (at least) of query hook instances on this page. All of the assets on the screen, links in the side bar, and virtually everything else, probably have some RQ hook nested somewhere in them. 

See this video showing more in-depth:

[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/xZGc8jr50s0/0.jpg)](https://www.youtube.com/watch?v=xZGc8jr50s0)

Does anyone have any ideas?

**Comment by TkDodo** [comment]
I'd be happy to look into this with you. I always advocate for just calling useQuery wherever you need the data, but of course creating an observer is not for free. However, it shouldn't be _that_ expensive I think.
Does it behave like that in production as well?
A codesandbox reproduction would be great - maybe just some dummy app where the same query is mounted a hundred times or so...

**Comment by dwilt** [comment]
Thanks for your offering to help. And yes, it does behave like this in production as well. If you sign up with our application at air.inc, and add a lot of assets to 2 boards, I think you'll be able to replicate this.

I'd love to do a screenshare with you. Can you find some time on [my calendar](https://calendly.com/dwilt-at-air)? Hopefully sooner than later 😅

**Reply by TkDodo** [reply]
do you have a workspace you can invite me to that shows this? I signed up as `dominik@dorfmeister.cc`

**Reply by dwilt** [reply]
> do you have a workspace you can invite me to that shows this? I signed up as `dominik@dorfmeister.cc`

Go sign up at qa.air.inc. This is our development server. I just sent you an invite to my workspace with these exact boards I'm showing in the video.

**Reply by dwilt** [reply]
I also just added you to my Slack so we could talk there if you'd like as well. We're debugging this at the moment

**Comment by dwilt** [comment]
@TkDodo I've put together [a CodeSandbox showing our issue](https://codesandbox.io/s/react-typescript-forked-8xw2bg?file=/src/MyRQComponent.tsx). When there are 100 items on the screen, with 8 observers in each item, at 6x slowdown, they take about a 1.2s to mount and about 750ms to unmount. You can see the entire thread is locked up for that time with all of the event listeners being added/removed.

![2022-06-28 09 43 11](https://user-images.githubusercontent.com/1065539/176194200-befc25d8-4752-4acb-b154-86ed75a6a7db.gif)

**Reply by TkDodo** [reply]
My initial profiling does indeed show that `clearTimeout` takes quite some time, and we call this for every observer.

see also this issue:
- https://github.com/highcharts/highcharts/issues/7901

we could do the same - just _not_ call `clearTimeout` if we didn't have a previous one. what do you think?

to confirm the speedup, I have set `staleTime: Infinity`, because there won't be a timer for infinite stale time. I have seen a significant speedup, but this is of course not something I would recommend for production.

However, this would likely not change a thing if you switch away from a view that has 800 observers, because there, we do need to clear the timeout ...

The real question is: why is clearTimeout so slow 😅

**Reply by tomaszczura** [reply]
Is it necessary to create `queryObserver` and these timeouts for each `useQuery` usage? Would it be possible to cache `queryObservers` by key/options or anything, and just add subscribers to each one?

**Reply by TkDodo** [reply]
observers and subscribers are the same thing. Each observer can have a different staleTime, so they run a different timer. We can't really batch this either because observers can unmount at different times, thus making the timer moot. A query is considered stale when at least one active observer's timer has elapsed.

So if the bottleneck is really `clearTimeout` - a browser native api - I'm not sure what I should do. In general, creating an observer does have a non-zero amount of overhead. The property tracking via `Object.defineProperty` also takes some time. The only other optimization that comes to my mind would be fewer observers. I'm guessing the 800 observers are not all individual queries. What might help is reducing them by creating one observer and then passing the relevant `data` down either as a prop, or via react context. That would avoid creating one observer at each leaf node.

**Reply by dwilt** [reply]
> but this is of course not something I would recommend for production.

Actually, we have a bunch of these `useQuery` instances that we could apply this to and it wouldn't have a negative impact. A lot of our endpoints don't change that much but we want the application to update if the _current_ user updates them. I'm assuming if we mutated a key with `Infinity`, the hook instances would still update, right?

**Reply by TkDodo** [reply]
> I'm assuming if we mutated a key with Infinity, it would still update, right?

It would still update if you _invalidate_ the key after the mutation, yes. A manual invalidation always overrides all timers and also Infinity staleTime.

**Reply by dwilt** [reply]
@tomaszczura is doing some testing to see how much better it would make our app. He's modifying the lib code in his local `node_modules` to test.

Probably something still that would be a good thing to add in, no? Why clear a timer if it doesn't exist?

**Reply by TkDodo** [reply]
> Probably something still that would be a good thing to add in, no? Why clear a timer if it doesn't exist?

yes, I would accept a PR that adds that fix :)

**Reply by TkDodo** [reply]
here's a quick benchmark showing that `clearTimeout(undefind)` slows everything down 5x:

https://jsbench.me/mjl4ybxzwe/1

**Reply by dwilt** [reply]
We're having a bit of trouble replicating the benefits on our side.. can you post up a codesandbox of the example with the improvement your seeing?

**Reply by TkDodo** [reply]
Hm, I'm using this sandbox: https://codesandbox.io/s/react-typescript-forked-67gdwl?file=/src/App.tsx

note that I set `initialData`, so the queryFn isn't really called. I did that to take re-renderings out of the equation from when data comes in, because that is something we cannot influence.

But we still call `clearTimeout` unnecessarily. The infinity staleTime would only help to avoid creating timers. So things should be better after we wrap the `clearTimeout` call in an `if` statement.

**Comment by dwilt** [comment, marked as answer]
**Update**: After talking with @TkDodo (thanks for your help), we think the best approach here is to just reduce the number of usages of `useQuery`. There are a bunch of instances of them that we can move up to the top of our application and only have one instance since they don't update very often. In short, using `useQuery` has a bit of overhead (like `useSWR`) and if you have a lot of elements with a lot of instances per element, this can seriously add up because of the way observers are added and removed on mount and unmount.

To show an example of our change, we previously had a component on our screen with a `useQuery` like this:

**Before:**
```tsx
// here's one of our reusable hooks that got used in a lot of places
const useSubscription = () => {
  const { currentWorkspace } = useCurrentWorkspace();
  const subscriptionKey = getSubscriptionKey(currentWorkspace?.id);
  
  return useQuery(subscriptionKey);
}

// here's how it was used and this component could be rendered dozens of times
// and in a virtualized list so improving the mount/unmount time is super important
const myComponent = () => {
  const { data: subscription } = useSubscription();
}
```

Since the user's subscription really never changed, and to save the perf, we moved the `useSubscription` hook to Context and now we don't have the issue of unmounting and remounting of observers:

**After:**

```tsx
export const SubscriptionProvider = ({ children }: SubscriptionProviderProps) => {
  // we don't destructure `isLoading` to reduce re-renders since Context API sucks at that
  const { data, error } = useSubscription();

  const value = useMemo(() => ({ data, error }), [data, error]);

  return <SubscriptionProviderContext.Provider value={value}>{children}</SubscriptionProviderContext.Provider>;
};

export const useSubscriptionContext = () => {
  const context = useContext(SubscriptionProviderContext);

  if (context === defaultValue) {
    const error = 'SubscriptionProviderContext used outside of SubscriptionProvider';

    if (isDevStage()) {
      throw error;
    } else {
      handleTryCatchError({
        error,
        context: error,
      });
    }
  }

  return context;
};

// and then in our component
const myComponent = () => {
  const { data: subscription } = useSubscriptionContext();
}
```
